# Mallory Drake, Luke Weeklund & Paige Turner Data Wrangling Project

## Cleaning & Merging the Datasets

In [3]:
#pip install rapidfuzz

In [4]:
# import needed packages
import pandas as pd
import matplotlib.pyplot as plt
from fuzzywuzzy import fuzz
from fuzzywuzzy import process
from rapidfuzz import process, fuzz

#### Upload the Kaggle dataset on University Rankings

In [6]:
# open kaggle excel file
university_rankings = pd.read_excel('2024 QS World University Rankings.xlsx',skiprows=[1])
university_rankings

2024 RANK  2023 RANK                              Institution Name  \
0            1          1  Massachusetts Institute of Technology (MIT)    
1            2          2                       University of Cambridge   
2            3          4                          University of Oxford   
3            4          5                            Harvard University   
4            5          3                           Stanford University   
...        ...        ...                                           ...   
1492     1401+      1401+                          University of Oradea   
1493     1401+        NaN                       University of Stavanger   
1494     1401+  1201-1400      University Politehnica of Timisoara, UPT   
1495     1401+  1201-1400                 Western Washington University   
1496     1401+        NaN                 Zürcher Hochschule Winterthur   

     Country Code         Country SIZE FOCUS RES.  AGE STATUS  ...  \
0              US   United States    M    CO   VH  5.0      B  ...   
1              UK  United Kingdom    L    FC   VH  5.0      A  ...   
2              UK  United Kingdom    L    FC   VH  5.0      A  ...   
3              US   United States    L    FC   VH  5.0      B  ...   
4              US   United States    L    FC   VH  5.0      B  ...   
...           ...             ...  ...   ...  ...  ...    ...  ...   
1492           RO         Romania    L    FC   MD  4.0      A  ...   
1493           NO          Norway  NaN    CO  NaN  2.0    NaN  ...   
1494           RO         Romania    M    FO   VH  5.0      A  ...   
1495           US   United States    L    CO   HI  5.0    NaN  ...   
1496           CH     Switzerland  NaN    CO  NaN  NaN    NaN  ...   

      International Faculty Rank International Students Score  \
0                             56                         88.2   
1                             64                         95.8   
2                            110                         98.2   
3                            210                         66.8   
4                             78                         51.2   
...                          ...                          ...   
1492                        701+                          7.0   
1493                         NaN                          NaN   
1494                        701+                          2.9   
1495                         676                          1.4   
1496                         NaN                          NaN   

      International Students Rank International Research Network Score  \
0                             128                                 94.3   
1                              85                                 99.9   
2                              60                                100.0   
3                             223                                100.0   
4                             284                                 95.8   
...                           ...                                  ...   
1492                         701+                                  1.8   
1493                          NaN                                 26.3   
1494                         701+                                  1.3   
1495                         701+                                  1.0   
1496                          NaN                                  1.7   

      International Research Network Rank Employment Outcomes Score  \
0                                      58                     100.0   
1                                       7                     100.0   
2                                       1                     100.0   
3                                       5                     100.0   
4                                      44                     100.0   
...                                   ...                       ...   
1492                                 701+                       5.9   
1493                     

##### convert the index to the 2024 rank value to fix the 1401+ issue and make them all numeric

In [8]:
university_rankings['2024 RANK'] = university_rankings.index + 1 
university_rankings

2024 RANK  2023 RANK                              Institution Name  \
0             1          1  Massachusetts Institute of Technology (MIT)    
1             2          2                       University of Cambridge   
2             3          4                          University of Oxford   
3             4          5                            Harvard University   
4             5          3                           Stanford University   
...         ...        ...                                           ...   
1492       1493      1401+                          University of Oradea   
1493       1494        NaN                       University of Stavanger   
1494       1495  1201-1400      University Politehnica of Timisoara, UPT   
1495       1496  1201-1400                 Western Washington University   
1496       1497        NaN                 Zürcher Hochschule Winterthur   

     Country Code         Country SIZE FOCUS RES.  AGE STATUS  ...  \
0              US   United States    M    CO   VH  5.0      B  ...   
1              UK  United Kingdom    L    FC   VH  5.0      A  ...   
2              UK  United Kingdom    L    FC   VH  5.0      A  ...   
3              US   United States    L    FC   VH  5.0      B  ...   
4              US   United States    L    FC   VH  5.0      B  ...   
...           ...             ...  ...   ...  ...  ...    ...  ...   
1492           RO         Romania    L    FC   MD  4.0      A  ...   
1493           NO          Norway  NaN    CO  NaN  2.0    NaN  ...   
1494           RO         Romania    M    FO   VH  5.0      A  ...   
1495           US   United States    L    CO   HI  5.0    NaN  ...   
1496           CH     Switzerland  NaN    CO  NaN  NaN    NaN  ...   

      International Faculty Rank International Students Score  \
0                             56                         88.2   
1                             64                         95.8   
2                            110                         98.2   
3                            210                         66.8   
4                             78                         51.2   
...                          ...                          ...   
1492                        701+                          7.0   
1493                         NaN                          NaN   
1494                        701+                          2.9   
1495                         676                          1.4   
1496                         NaN                          NaN   

      International Students Rank International Research Network Score  \
0                             128                                 94.3   
1                              85                                 99.9   
2                              60                                100.0   
3                             223                                100.0   
4                             284                                 95.8   
...                           ...                                  ...   
1492                         701+                                  1.8   
1493                          NaN                                 26.3   
1494                         701+                                  1.3   
1495                         701+                                  1.0   
1496                          NaN                                  1.7   

      International Research Network Rank Employment Outcomes Score  \
0                                      58                     100.0   
1                                       7                     100.0   
2                                       1                     100.0   
3                                       5                     100.0   
4                                      44                     100.0   
...                                   ...                       ...   
1492                                 701+                       5.9   
1493          

#### Clean the Kaggle Dataset to remove any abbreviations in parentheses from the Institution name to match the school name format in the scraped table

In [10]:
university_rankings['Institution Name'] = university_rankings['Institution Name'].str.replace(r'\(.*?\)', '', regex=True).str.strip()
university_rankings

2024 RANK  2023 RANK                          Institution Name  \
0             1          1     Massachusetts Institute of Technology   
1             2          2                   University of Cambridge   
2             3          4                      University of Oxford   
3             4          5                        Harvard University   
4             5          3                       Stanford University   
...         ...        ...                                       ...   
1492       1493      1401+                      University of Oradea   
1493       1494        NaN                   University of Stavanger   
1494       1495  1201-1400  University Politehnica of Timisoara, UPT   
1495       1496  1201-1400             Western Washington University   
1496       1497        NaN             Zürcher Hochschule Winterthur   

     Country Code         Country SIZE FOCUS RES.  AGE STATUS  ...  \
0              US   United States    M    CO   VH  5.0      B  ...   
1              UK  United Kingdom    L    FC   VH  5.0      A  ...   
2              UK  United Kingdom    L    FC   VH  5.0      A  ...   
3              US   United States    L    FC   VH  5.0      B  ...   
4              US   United States    L    FC   VH  5.0      B  ...   
...           ...             ...  ...   ...  ...  ...    ...  ...   
1492           RO         Romania    L    FC   MD  4.0      A  ...   
1493           NO          Norway  NaN    CO  NaN  2.0    NaN  ...   
1494           RO         Romania    M    FO   VH  5.0      A  ...   
1495           US   United States    L    CO   HI  5.0    NaN  ...   
1496           CH     Switzerland  NaN    CO  NaN  NaN    NaN  ...   

      International Faculty Rank International Students Score  \
0                             56                         88.2   
1                             64                         95.8   
2                            110                         98.2   
3                            210                         66.8   
4                             78                         51.2   
...                          ...                          ...   
1492                        701+                          7.0   
1493                         NaN                          NaN   
1494                        701+                          2.9   
1495                         676                          1.4   
1496                         NaN                          NaN   

      International Students Rank International Research Network Score  \
0                             128                                 94.3   
1                              85                                 99.9   
2                              60                                100.0   
3                             223                                100.0   
4                             284                                 95.8   
...                           ...                                  ...   
1492                         701+                                  1.8   
1493                          NaN                                 26.3   
1494                         701+                                  1.3   
1495                         701+                                  1.0   
1496                          NaN                                  1.7   

      International Research Network Rank Employment Outcomes Score  \
0                                      58                     100.0   
1                                       7                     100.0   
2                                       1                     100.0   
3                                       5                     100.0   
4                                      44                     100.0   
...                                   ...                       ...   
1492                                 701+                       5.9   
1493                                  489                     

#### Open the scraped dataset csv file

In [12]:
scraped_salary_data = pd.read_csv('scraped_univeristy_salary_data.csv')
scraped_salary_data

Rank                            School Name  \
0        1  Massachusetts Institute of Technology   
1        2                   Princeton University   
2        3            United States Naval Academy   
3        4                    Harvey Mudd College   
4        5                         Babson College   
...    ...                                    ...   
1498  1499                         Morris College   
1499  1500                          Texas College   
1500  1501                     Claflin University   
1501  1501         The Baptist College of Florida   
1502  1503          Great Lakes Christian College   

                                            School Type Early Career Pay  \
0      Engineering, Private School, Research University        $110,200    
1     Engineering, Ivy League, Private School, Resea...         $95,600    
2     Engineering, Liberal Arts School, Sober School...         $96,700    
3      Engineering, Liberal Arts School, Private School        $115,000    
4                              Business, Private School         $90,600    
...                                                 ...              ...   
1498                          Private School, Religious         $40,000    
1499                          Private School, Religious         $42,700    
1500     Liberal Arts School, Private School, Religious         $47,900    
1501                          Private School, Religious         $41,900    
1502                          Private School, Religious         $39,500    

     Mid Career Pay Percent of STEM Degrees  
0         $196,900                      66%  
1         $194,100                      53%  
2         $187,800                      54%  
3         $185,900                      74%  
4         $181,400                       3%  
...             ...                     ...  
1498       $68,600                       1%  
1499       $68,300                      21%  
1500       $68,000                      22%  
1501       $68,000                       0%  
1502       $65,200                       0%  

[1503 rows x 6 columns]

#### Clean the Kaggle dataset

In [14]:
# drop unnecessary columns (2023 rank, country code, focus)
university_rankings = university_rankings.drop(['2023 RANK', 'Country Code', 'FOCUS', 'Citations per Faculty Score', 'Citations per Faculty Rank', 'International Faculty Rank', 'International Faculty Score', 'International Students Score', 'International Students Rank', 'International Research Network Score', 'International Research Network Rank', 'RES.', 'STATUS'], axis=1)
university_rankings.head()


2024 RANK                       Institution Name         Country SIZE  AGE  \
0          1  Massachusetts Institute of Technology   United States    M  5.0   
1          2                University of Cambridge  United Kingdom    L  5.0   
2          3                   University of Oxford  United Kingdom    L  5.0   
3          4                     Harvard University   United States    L  5.0   
4          5                    Stanford University   United States    L  5.0   

   Academic Reputation Score Academic Reputation Rank  \
0                      100.0                        4   
1                      100.0                        3   
2                      100.0                        2   
3                      100.0                        1   
4                      100.0                        5   

   Employer Reputation Score Employer Reputation Rank  Faculty Student Score  \
0                      100.0                        2                  100.0   
1                      100.0                        4                  100.0   
2                      100.0                        5                  100.0   
3                      100.0                        1                   98.3   
4                      100.0                        3                  100.0   

  Faculty Student Rank  Employment Outcomes Score Employment Outcomes Rank  \
0                   16                      100.0                        4   
1                   14                      100.0                        6   
2                    8                      100.0                        3   
3                   44                      100.0                        1   
4                   11                      100.0                        2   

   Sustainability Score Sustainability Rank Overall SCORE  
0                  95.2                  51           100  
1                  97.3                 33=          99.2  
2                  97.8                 26=          98.9  
3                  96.7                  39          98.3  
4                  94.4                  63          98.1

In [15]:
# rename the Overall SCORE, RES., and 2024 RANK columns
university_rankings.rename(columns={'Overall SCORE': 'Overall Score', '2024 RANK':'2024 Rank'}, inplace=True)

In [16]:
# rename the institution name column in the university rankings dataframe to school name
university_rankings.rename(columns={'Institution Name': 'School Name'}, inplace=True)
university_rankings.head()

2024 Rank                            School Name         Country SIZE  AGE  \
0          1  Massachusetts Institute of Technology   United States    M  5.0   
1          2                University of Cambridge  United Kingdom    L  5.0   
2          3                   University of Oxford  United Kingdom    L  5.0   
3          4                     Harvard University   United States    L  5.0   
4          5                    Stanford University   United States    L  5.0   

   Academic Reputation Score Academic Reputation Rank  \
0                      100.0                        4   
1                      100.0                        3   
2                      100.0                        2   
3                      100.0                        1   
4                      100.0                        5   

   Employer Reputation Score Employer Reputation Rank  Faculty Student Score  \
0                      100.0                        2                  100.0   
1                      100.0                        4                  100.0   
2                      100.0                        5                  100.0   
3                      100.0                        1                   98.3   
4                      100.0                        3                  100.0   

  Faculty Student Rank  Employment Outcomes Score Employment Outcomes Rank  \
0                   16                      100.0                        4   
1                   14                      100.0                        6   
2                    8                      100.0                        3   
3                   44                      100.0                        1   
4                   11                      100.0                        2   

   Sustainability Score Sustainability Rank Overall Score  
0                  95.2                  51           100  
1                  97.3                 33=          99.2  
2                  97.8                 26=          98.9  
3                  96.7                  39          98.3  
4                  94.4                  63          98.1

##### Map the school names that are similar but not the same in the 2 datasets
        work was done manually in a spreadsheet because the rapidfuzz techniques below didn't work.

In [18]:
# create a mapping dictionary between School Name in the Kaggle dataset and the scraped data set 
school_name_mapping = {
    'Arizona State University': 'Arizona State University Campus Immersion',
    'Binghamton University SUNY': 'Binghamton University',
    'California Polytechnic State University': 'California Polytechnic State University-San Luis Obispo',
    'California State University Long Beach': 'California State University-Long Beach',
    'California State University - Los Angeles': 'California State University-Los Angeles',
    'Colorado State University': 'Colorado State University-Fort Collins',
    'Columbia University': 'Columbia University in the City of New York',
    'CUNY The City College of New York': 'CUNY City College',
    'Florida Atlantic University - Boca Raton': 'Florida Atlantic University',
    'Georgia Institute of Technology': 'Georgia Institute of Technology-Main Campus',
    'Indiana University Bloomington': 'Indiana University-Bloomington',
    'Indiana University–Purdue University Indianapolis': 'Indiana University-Purdue University-Indianapolis',
    'Kent State University': 'Kent State University at Kent',
    "King's College London": "King's College",
    'Lincoln University': 'Lincoln University (Pennsylvania)',
    'Miami University': 'Miami University-Oxford',
    'North Carolina State University': 'North Carolina State University at Raleigh',
    'The Ohio State University': 'Ohio State University-Main Campus',
    'Ohio University': 'Ohio University-Main Campus',
    'Oklahoma State University': 'Oklahoma State University-Main Campus',
    'Purdue University': 'Purdue University-Main Campus',
    'Rutgers, The State University of New Jersey, Camden': 'Rutgers University-Camden',
    'Rutgers University–New Brunswick': 'Rutgers University-New Brunswick',
    'Rutgers University–Newark': 'Rutgers University-Newark',
    'Texas A&M University': 'Texas A & M University-College Station',
    'The University of Tennessee, Knoxville': 'The University of Tennessee-Knoxville',
    'University at Albany SUNY': 'SUNY at Albany',
    'University at Buffalo SUNY': 'University at Buffalo',
    'University of Alabama at Birmingham': 'University of Alabama at Birmingham',
    'The University of Arizona': 'University of Arizona',
    'University of Arkansas Fayetteville': 'University of Arkansas',
    'University of California, Berkeley': 'University of California-Berkeley',
    'University of California, Davis': 'University of California-Davis',
    'University of California, Irvine': 'University of California-Irvine',
    'University of California, Los Angeles': 'University of California-Los Angeles',
    'University of California, Riverside': 'University of California-Riverside',
    'University of California, San Diego': 'University of California-San Diego',
    'University of California, Santa Barbara': 'University of California-Santa Barbara',
    'University of California, Santa Cruz': 'University of California-Santa Cruz',
    'University of Cincinnati': 'University of Cincinnati-Main Campus',
    'University of Colorado, Denver': 'University of Colorado Denver/Anschutz Medical Campus',
    'The University of Georgia': 'University of Georgia',
    'University of Hawaiʻi at Mānoa': 'University of Hawaii at Manoa',
    'University of Illinois at Chicago': 'University of Illinois Chicago',
    'University of Illinois at Urbana-Champaign': 'University of Illinois Urbana-Champaign',
    'University of Maryland, Baltimore County': 'University of Maryland-Baltimore County',
    'University of Maryland, College Park': 'University of Maryland-College Park',
    'University of Massachusetts Amherst': 'University of Massachusetts-Amherst',
    'University of Massachusetts Boston': 'University of Massachusetts-Boston',
    'University of Minnesota Twin Cities': 'University of Minnesota-Twin Cities',
    'University of Missouri Saint Louis': 'University of Missouri-St Louis',
    'University of Missouri, Columbia': 'University of Missouri-Columbia',
    'University of Missouri, Kansas City': 'University of Missouri-Kansas City',
    'University of Nebraska - Lincoln': 'University of Nebraska-Lincoln',
    'University of New Mexico': 'University of New Mexico-Main Campus',
    'University of North Carolina, Chapel Hill': 'University of North Carolina at Chapel Hill',
    'University of Oklahoma': 'University of Oklahoma-Norman Campus',
    'University of Pittsburgh': 'University of Pittsburgh-Pittsburgh Campus',
    'University of South Carolina': 'University of South Carolina-Columbia',
    'University of Virginia': 'University of Virginia-Main Campus',
    'University of Washington': 'University of Washington-Seattle Campus',
    'University of Wisconsin Milwaukee': 'University of Wisconsin-Milwaukee'
}

In [19]:
# Apply the mapping to standardize names
university_rankings['School Name'] = university_rankings['School Name'].map(school_name_mapping).fillna(university_rankings['School Name'])

#### hyperparameter testing that was unsuccessful

In [21]:
"""
# Initialize an empty list to store the closest matches
match_counts = []

# Try different matching thresholds from 65 to 95
thresholds = range(65, 95, 2)  # This will test thresholds from 65, 67, ..., 93

# Perform fuzzy matching for each threshold
for threshold in thresholds:
    matches = []

    for school in university_rankings['School Name']:
        match = process.extractOne(school, scraped_salary_data['School Name'], scorer=fuzz.ratio)
        if match and match[1] >= threshold:  # Only keep matches above the threshold
            matches.append((school, match[0], match[1]))

    # Count the number of matches for this threshold
    match_counts.append(len(matches))

# Plotting the results
plt.figure(figsize=(10, 6))
plt.plot(thresholds, match_counts, marker='o', linestyle='-', color='b')
plt.title("Number of Matches at Different Match Thresholds", fontsize=16)
plt.xlabel("Match Score Threshold", fontsize=12)
plt.ylabel("Number of Matches", fontsize=12)
plt.grid(True)
plt.show()

# Output the number of matches for each threshold
match_results = pd.DataFrame({
    'Threshold': thresholds,
    'Number of Matches': match_counts
})

match_results
"""

'\n# Initialize an empty list to store the closest matches\nmatch_counts = []\n\n# Try different matching thresholds from 65 to 95\nthresholds = range(65, 95, 2)  # This will test thresholds from 65, 67, ..., 93\n\n# Perform fuzzy matching for each threshold\nfor threshold in thresholds:\n    matches = []\n\n    for school in university_rankings[\'School Name\']:\n        match = process.extractOne(school, scraped_salary_data[\'School Name\'], scorer=fuzz.ratio)\n        if match and match[1] >= threshold:  # Only keep matches above the threshold\n            matches.append((school, match[0], match[1]))\n\n    # Count the number of matches for this threshold\n    match_counts.append(len(matches))\n\n# Plotting the results\nplt.figure(figsize=(10, 6))\nplt.plot(thresholds, match_counts, marker=\'o\', linestyle=\'-\', color=\'b\')\nplt.title("Number of Matches at Different Match Thresholds", fontsize=16)\nplt.xlabel("Match Score Threshold", fontsize=12)\nplt.ylabel("Number of Matches", f

##### use rapid fuzz to find best matches between school name columns - unsuccessful

In [23]:
"""
# Create a mapping of the closest matches
matched_names = []

for school in university_rankings['School Name']:
    match, score, _ = process.extractOne(
        school,
        scraped_salary_data['School Name'],
        scorer=fuzz.token_sort_ratio
    )
    
    # Set a threshold for match quality
    if score >= 85:
        matched_names.append({'Original': school, 'Matched': match, 'Score': score})

# Convert the matched names to a DataFrame
matches_df = pd.DataFrame(matched_names)

# Merge the match DataFrame with university_rankings to get the matched names
university_rankings_matched = university_rankings.merge(
    matches_df,
    left_on='School Name',
    right_on='Original',
    how='left'
)

# Final merge with the salary data using the matched names
final_merged = university_rankings_matched.merge(
    scraped_salary_data,
    left_on='Matched',
    right_on='School Name',
    suffixes=('_ranking', '_salary'),
    how='inner'  # or 'left' if you want to keep all from university_rankings
)

# Optional: Drop duplicate columns or clean up as needed
final_merged.drop(columns=['Original', 'Matched', 'Score'], inplace=True)

final_merged
"""

"\n# Create a mapping of the closest matches\nmatched_names = []\n\nfor school in university_rankings['School Name']:\n    match, score, _ = process.extractOne(\n        school,\n        scraped_salary_data['School Name'],\n        scorer=fuzz.token_sort_ratio\n    )\n    \n    # Set a threshold for match quality\n    if score >= 85:\n        matched_names.append({'Original': school, 'Matched': match, 'Score': score})\n\n# Convert the matched names to a DataFrame\nmatches_df = pd.DataFrame(matched_names)\n\n# Merge the match DataFrame with university_rankings to get the matched names\nuniversity_rankings_matched = university_rankings.merge(\n    matches_df,\n    left_on='School Name',\n    right_on='Original',\n    how='left'\n)\n\n# Final merge with the salary data using the matched names\nfinal_merged = university_rankings_matched.merge(\n    scraped_salary_data,\n    left_on='Matched',\n    right_on='School Name',\n    suffixes=('_ranking', '_salary'),\n    how='inner'  # or 'left' 

##### use fuzzy wuzzy program to find best matches between school name columns - unsuccessful

In [25]:
"""
# Initialize an empty list to store the closest matches
matches = []

# Perform fuzzy matching for each School Name in university_rankings
for school in university_rankings['School Name']:
    match = process.extractOne(school, scraped_salary_data['School Name'], scorer=fuzz.ratio)
    if match and match[1] >= 95:  # Only keep matches with a score >= 90
        matches.append((school, match[0], match[1]))

# Create a DataFrame from the matches
match_df = pd.DataFrame(matches, columns=['School Name (QS)', 'School Name (Salary Data)', 'Match Score'])

# Merge the datasets based on the closest fuzzy matches
merged_university_rankings = pd.merge(university_rankings, match_df, left_on='School Name', right_on='School Name (QS)', how='inner')

# Merge with scraped salary data based on fuzzy matched school names
merged_university_rankings = pd.merge(merged_university_rankings, scraped_salary_data, left_on='School Name (Salary Data)', right_on='School Name', how='inner')

# Drop extra columns used for merging
merged_university_rankings.drop(columns=['School Name (Salary Data)', 'School Name (QS)'], inplace=True)

# Display the merged DataFrame
merged_university_rankings
"""

"\n# Initialize an empty list to store the closest matches\nmatches = []\n\n# Perform fuzzy matching for each School Name in university_rankings\nfor school in university_rankings['School Name']:\n    match = process.extractOne(school, scraped_salary_data['School Name'], scorer=fuzz.ratio)\n    if match and match[1] >= 95:  # Only keep matches with a score >= 90\n        matches.append((school, match[0], match[1]))\n\n# Create a DataFrame from the matches\nmatch_df = pd.DataFrame(matches, columns=['School Name (QS)', 'School Name (Salary Data)', 'Match Score'])\n\n# Merge the datasets based on the closest fuzzy matches\nmerged_university_rankings = pd.merge(university_rankings, match_df, left_on='School Name', right_on='School Name (QS)', how='inner')\n\n# Merge with scraped salary data based on fuzzy matched school names\nmerged_university_rankings = pd.merge(merged_university_rankings, scraped_salary_data, left_on='School Name (Salary Data)', right_on='School Name', how='inner')\n\n#

#### Merge 2 dataframes using an inner join on School Name 

In [27]:
merged_university_rankings = pd.merge(university_rankings, scraped_salary_data, on='School Name', how='inner')
merged_university_rankings

2024 Rank                              School Name        Country SIZE  \
0            1    Massachusetts Institute of Technology  United States    M   
1            4                       Harvard University  United States    L   
2            5                      Stanford University  United States    L   
3           10        University of California-Berkeley  United States   XL   
4           11                    University of Chicago  United States    L   
..         ...                                      ...            ...  ...   
177       1402                University of the Pacific  United States    M   
178       1415  California State University-Los Angeles  United States    L   
179       1416   California State University-Long Beach  United States   XL   
180       1421                Illinois State University  United States    L   
181       1496            Western Washington University  United States    L   

     AGE  Academic Reputation Score Academic Reputation Rank  \
0    5.0                      100.0                        4   
1    5.0                      100.0                        1   
2    5.0                      100.0                        5   
3    5.0                      100.0                        6   
4    5.0                       99.1                       17   
..   ...                        ...                      ...   
177  5.0                        2.2                     601+   
178  4.0                        5.4                     601+   
179  4.0                        4.0                     601+   
180  5.0                        3.7                     601+   
181  5.0                        2.8                     601+   

     Employer Reputation Score Employer Reputation Rank  \
0                        100.0                        2   
1                        100.0                        1   
2                        100.0                        3   
3                        100.0                        6   
4                         95.4                       38   
..                         ...                      ...   
177                        3.7                     601+   
178                        2.5                     601+   
179                        2.8                     601+   
180                        3.5                     601+   
181                        3.4                     601+   

     Faculty Student Score  ... Employment Outcomes Score  \
0                    100.0  ...                     100.0   
1                     98.3  ...                     100.0   
2                    100.0  ...                     100.0   
3                     20.5  ...                      98.7   
4                     92.5  ...                      98.3   
..                     ...  ...                       ...   
177                   34.3  ...                      10.8   
178                    3.5  ...                       9.5   
179                    8.1  ...                       9.6   
180                    5.9  ...                      10.9   
181                    7.0  ...                       7.3   

     Employment Outcomes Rank Sustainability Score  Sustainability Rank  \
0                           4                 95.2                   51   
1                           1                 96.7                   39   
2                           2                 94.4                   63   
3                          15                100.0                    1   
4                          17                 81.7                 137=   
..                        ...                  ...                  ...   
177                      701+                  1.0                 701+   
178                      701+                  NaN                  NaN   
179                      701+                  NaN                  NaN   
180                      701+                  NaN                  NaN   
181                 

In [28]:
# change scraped rank to post grad salary rank
merged_university_rankings.rename(columns={'Rank': 'Postgrad Salary Rank'}, inplace=True)

In [29]:
# drop school type and all the individual ranks
merged_university_rankings = merged_university_rankings.drop(['School Type', 'Academic Reputation Rank', 'Employer Reputation Rank', 'Faculty Student Rank', 'Employment Outcomes Rank', 'Sustainability Rank', 'Sustainability Score'], axis=1)

#### Change the Numeric Data Columns to the Correct Numeric Data Type

In [31]:
merged_university_rankings.dtypes

2024 Rank                      int64
School Name                   object
Country                       object
SIZE                          object
AGE                          float64
Academic Reputation Score    float64
Employer Reputation Score    float64
Faculty Student Score        float64
Employment Outcomes Score    float64
Overall Score                 object
Postgrad Salary Rank           int64
Early Career Pay              object
Mid Career Pay                object
Percent of STEM Degrees       object
dtype: object

In [38]:
# Convert numeric columns to numeric type
numeric_columns = [
    '2024 Rank', 'AGE', 'Academic Reputation Score', 
    'Employer Reputation Score', 
    'Faculty Student Score', 
    'Employment Outcomes Score', 
    'Overall Score', 'Postgrad Salary Rank'
]
for column in numeric_columns:
    merged_university_rankings[column] = pd.to_numeric(merged_university_rankings[column], errors='coerce')

merged_university_rankings.dtypes

2024 Rank                      int64
School Name                   object
Country                       object
SIZE                          object
AGE                          float64
Academic Reputation Score    float64
Employer Reputation Score    float64
Faculty Student Score        float64
Employment Outcomes Score    float64
Overall Score                float64
Postgrad Salary Rank           int64
Early Career Pay              object
Mid Career Pay                object
Percent of STEM Degrees       object
dtype: object

In [40]:
# remove text characters from numeric columns and converting it to numeric
merged_university_rankings['Early Career Pay'] = merged_university_rankings['Early Career Pay'].str.replace(r'[\$,]', '', regex=True).astype(float)
merged_university_rankings['Mid Career Pay'] = merged_university_rankings['Mid Career Pay'].str.replace(r'[\$,]', '', regex=True).astype(float)
merged_university_rankings['Percent of STEM Degrees'] = merged_university_rankings['Percent of STEM Degrees'].str.replace('%', '').astype(float)


In [42]:
# check that all the new data types are correct
merged_university_rankings.dtypes

2024 Rank                      int64
School Name                   object
Country                       object
SIZE                          object
AGE                          float64
Academic Reputation Score    float64
Employer Reputation Score    float64
Faculty Student Score        float64
Employment Outcomes Score    float64
Overall Score                float64
Postgrad Salary Rank           int64
Early Career Pay             float64
Mid Career Pay               float64
Percent of STEM Degrees      float64
dtype: object

In [44]:
merged_university_rankings

2024 Rank                              School Name        Country SIZE  \
0            1    Massachusetts Institute of Technology  United States    M   
1            4                       Harvard University  United States    L   
2            5                      Stanford University  United States    L   
3           10        University of California-Berkeley  United States   XL   
4           11                    University of Chicago  United States    L   
..         ...                                      ...            ...  ...   
177       1402                University of the Pacific  United States    M   
178       1415  California State University-Los Angeles  United States    L   
179       1416   California State University-Long Beach  United States   XL   
180       1421                Illinois State University  United States    L   
181       1496            Western Washington University  United States    L   

     AGE  Academic Reputation Score  Employer Reputation Score  \
0    5.0                      100.0                      100.0   
1    5.0                      100.0                      100.0   
2    5.0                      100.0                      100.0   
3    5.0                      100.0                      100.0   
4    5.0                       99.1                       95.4   
..   ...                        ...                        ...   
177  5.0                        2.2                        3.7   
178  4.0                        5.4                        2.5   
179  4.0                        4.0                        2.8   
180  5.0                        3.7                        3.5   
181  5.0                        2.8                        3.4   

     Faculty Student Score  Employment Outcomes Score  Overall Score  \
0                    100.0                      100.0          100.0   
1                     98.3                      100.0           98.3   
2                    100.0                      100.0           98.1   
3                     20.5                       98.7           90.4   
4                     92.5                       98.3           90.3   
..                     ...                        ...            ...   
177                   34.3                       10.8            NaN   
178                    3.5                        9.5            NaN   
179                    8.1                        9.6            NaN   
180                    5.9                       10.9            NaN   
181                    7.0                        7.3            NaN   

     Postgrad Salary Rank  Early Career Pay  Mid Career Pay  \
0                       1          110200.0        196900.0   
1                      10           95600.0        177400.0   
2                       6          102300.0        181200.0   
3                      20           90900.0        170100.0   
4                      61           84700.0        151000.0   
..                    ...               ...             ...   
177                   161           75900.0        134800.0   
178                   471           62600.0        114400.0   
179                   258           67300.0        125000.0   
180                   425           64000.0        116300.0   
181                   231           70600.0        127500.0   

     Percent of STEM Degrees  
0                       66.0  
1                       19.0  
2                       48.0  
3                       44.0  
4                       30.0  
..                       ...  
177                     20.0  
178                     14.0  
179                     17.0  
180                     10.0  
181                     17.0  

[182 rows x 14 columns]

In [46]:
# save merged data to csv 
merged_university_rankings.to_csv('merged&cleaned_university_rankings.csv', index=False)